File: src/main/java/com/gtladd/gtladditions/api/machine/GTLAddCoilWorkableElectricMultipleRecipesMultiblockMachine.java
package com.gtladd.gtladditions.api.machine;

import org.gtlcore.gtlcore.api.machine.multiblock.ParallelMachine;

import com.gregtechceu.gtceu.api.machine.IMachineBlockEntity;
import com.gregtechceu.gtceu.api.machine.multiblock.CoilWorkableElectricMultiblockMachine;
import com.gregtechceu.gtceu.api.machine.trait.RecipeLogic;

import com.gtladd.gtladditions.api.recipeslogic.GTLAddMultipleRecipesLogic;
import org.jetbrains.annotations.NotNull;

public class GTLAddCoilWorkableElectricMultipleRecipesMultiblockMachine extends CoilWorkableElectricMultiblockMachine implements ParallelMachine {

    public GTLAddCoilWorkableElectricMultipleRecipesMultiblockMachine(IMachineBlockEntity holder) {
        super(holder);
    }

    public @NotNull RecipeLogic createRecipeLogic(@NotNull Object... args) {
        return new GTLAddMultipleRecipesLogic(this);
    }

    public @NotNull GTLAddMultipleRecipesLogic getRecipeLogic() {
        return (GTLAddMultipleRecipesLogic) super.getRecipeLogic();
    }

    @Override
    public int getMaxParallel() {
        return Math.min(Integer.MAX_VALUE, (int) Math.pow(2.0, (double) this.getCoilType().getCoilTemperature() / 900.0));
    }
}

File: src/main/java/com/gtladd/gtladditions/api/machine/GTLAddCoilWorkableElectricMultipleRecipesMultiblockMachine.kt
package com.gtladd.gtladditions.api.machine

import com.gregtechceu.gtceu.api.block.ICoilType
import com.gregtechceu.gtceu.api.machine.IMachineBlockEntity
import com.gregtechceu.gtceu.api.machine.trait.RecipeLogic
import com.gregtechceu.gtceu.common.block.CoilBlock
import com.gtladd.gtladditions.api.machine.logic.GTLAddMultipleRecipesLogic
import kotlin.math.min
import kotlin.math.pow

open class GTLAddCoilWorkableElectricMultipleRecipesMultiblockMachine(holder: IMachineBlockEntity) :
    GTLAddWorkableElectricMultipleRecipesMachine(holder) {
        var coilType: ICoilType = CoilBlock.CoilType.CUPRONICKEL

    override fun createRecipeLogic(vararg args: Any): RecipeLogic {
        return GTLAddMultipleRecipesLogic(this, EBFChecks.EBF_CHECK)
    }

    override fun getMaxParallel(): Int {
        return min(Int.Companion.MAX_VALUE, 2.0.pow(this.coilType.coilTemperature.toDouble() / 900.0).toInt())
    }

    override fun onStructureFormed() {
        super.onStructureFormed()
        val type = multiblockState.matchContext.get<Any?>("CoilType")
        if (type is ICoilType) this.coilType = type
    }

    fun getCoilTier(): Int {
        return coilType.tier
    }
}
